RoslynPluginGenerator/ArchiveUpdater.cs

```diff
@@ -0,0 +1,154 @@
+        }
+
+        public ArchiveUpdater AddFile(string sourceFilePath, string relativeTargetFilePath)
+        {
+            this.fileMap[relativeTargetFilePath] = sourceFilePath;
+            return this;
+        }
+
+        public void UpdateArchive()
+        {
+            this.logger.LogDebug(UIResources.ZIP_UpdatingArchive, this.inputArchiveFilePath);
+
+            string unpackedDir = Utilities.CreateSubDirectory(this.workingDirectory, "unpacked");
+            this.logger.LogDebug(UIResources.ZIP_WorkingDirectory, workingDirectory);
+
+            ZipFile.ExtractToDirectory(this.inputArchiveFilePath, unpackedDir);
+
+            // Add in the new files
+            foreach (KeyValuePair<string, string> kvp in this.fileMap)
+            {
+                this.logger.LogDebug(UIResources.ZIP_InsertingFile, kvp.Key, kvp.Value);
+
+                string targetFilePath = Path.Combine(unpackedDir, kvp.Key);
+                Directory.CreateDirectory(Path.GetDirectoryName(targetFilePath));
+                File.Copy(kvp.Value, targetFilePath);
+            }
+
+            // Re-zip
+            if (File.Exists(this.outputArchiveFilePath))
+            {
+                this.logger.LogDebug(UIResources.ZIP_DeletingExistingArchive);
+                File.Delete(this.outputArchiveFilePath);
+            }
+
+            ZipUsingShell(unpackedDir, this.outputArchiveFilePath);
+
+            this.logger.LogDebug(UIResources.ZIP_NewArchiveCreated, this.outputArchiveFilePath);
+        }
+        
+        private static void ZipUsingShell(string sourceDir, string targetZipFilePath)
+        {
+            // The Folder.CopyHere method for Shell Objects allows configuration based on a combination of flags.
+            // Docs here: https://msdn.microsoft.com/en-us/library/windows/desktop/bb787866(v=vs.85).aspx
+            // The value below (1556) consists of
+            //    (4)    - no progress dialog
+            //    (16)   - respond with "yes to all" to any dialog box
+            //    (512)  - Do not confirm the creation of a new directory
+            //    (1024) - Do not display an UI in case of error
+            const int copyFlags = 1556;
+
+            const int copyPauseInMilliseconds = 500;
+
+            // The file must have a ".zip" extension for the shell code below to work
+            string zipFilePath = targetZipFilePath + ".zip";
+            CreateEmptyZipFile(zipFilePath);
+
+            Type shellAppType = Type.GetTypeFromProgID("Shell.Application");
+            Object app = Activator.CreateInstance(shellAppType);
+
+            Shell32.Folder folder = shellAppType.InvokeMember("NameSpace", System.Reflection.BindingFlags.InvokeMethod, null, app, new object[] { zipFilePath }) as Shell32.Folder;
+
+            foreach(string dir in Directory.GetDirectories(sourceDir))
+            {
+                folder.CopyHere(dir, copyFlags);
+                System.Threading.Thread.Sleep(copyPauseInMilliseconds);
```

Comment:
It's not, but the copy is done on a different thread and we have no way of knowing that it has finished.